### PR TITLE
Automatic Documentation Builds & Actions Modifications

### PR DESCRIPTION
**Pull request by martin-sicho**
Added a workflow to automatically build and deploy documentation pages (`docs_ci.yml`). It is invoked upon push to `main` or `dev`. If a development version is detected, the documentation pages are deployed under `/docs-dev`, otherwise they will be deployed under `/docs`.

In addition, modified the `pre-merge-ci.yml` workflow to run on push to all branches, but `dev`, `main` and `gh_pages`. This workflow only runs simple tests, leaving out tutorial testing and testing of the `qsprpred.extra` package. The full `dev_ci.yml` workflow is triggered only when a merge request or push is made to either `dev` or `main`.

Finally, the CI workflow was also refactored moving some portions of the workflow to scripts.

### CODE REVIEW

File: qsprpred/data/tables/mol.py
Comment (martin-sicho) [inline qsprpred/data/tables/mol.py:1]: Also, a small fix that caused Pandas 2.2 to throw an exception was introduced here. This should be mentioned in the final change log.

File: qsprpred/extra/data/descriptors/sets.py [hunk 1]
```diff
@@ -194,35 +194,56 @@ class PaDEL(DescriptorSet):
...
                 continue
             for name in descriptor.description.name:
                 self._nameMapping[name] = descriptor
-        # Initialize descriptors and calculator
-        if descs is None:
-            self.descriptors = None
-        else:
-            self.descriptors = descs
+
+    def __setstate__(self, state):
```
Comment (martin-sicho) [inline qsprpred/extra/data/descriptors/sets.py:237]: We had a serialization issue here  with some members not being directly serializable anymore. This was fixed by leaving out some members  in  `_notJSON` and then re-initializing  them in `__setstate__`.

File: qsprpred/extra/data/descriptors/sets.py [hunk 2]
```diff
@@ -194,35 +194,56 @@ class PaDEL(DescriptorSet):
...
+        self._initMapping()
+        # Initialize descriptors and calculator
+        if descs is None:
+            self.descriptors = None
+        else:
+            self.descriptors = descs
+
+    @property
+    def supportsParallel(self) -> bool:
+        return False
```
Comment (martin-sicho) [inline qsprpred/extra/data/descriptors/sets.py:225]: Also labelled this processor as serial execution only, but the `nJobs` attribute  was added that can use the default parallelization from the PaDEL package.

**Comment by martin-sicho** [comment]
The Linux test passes so lets consider this done. @David-Araripe, please, have a look. I will address the MacOS issue in a separate PR later.

File: .github/workflows/dev_ci.yml [hunk 3]
```diff
@@ -1,82 +1,70 @@
...
 jobs:
-  build-and-test:
+  tests-all:
     runs-on: ${{ matrix.os }}
+    timeout-minutes: 120
     strategy:
       matrix:
-        os: [ubuntu-latest, macos-latest]
-        python-version: ['3.10']
+        os: [ ubuntu-latest ] # FIXME: add macos-latest and windows-latest (removed for now due to limited debugging options)
```
Comment (martin-sicho) [inline .github/workflows/dev_ci.yml:23]: In the end, I added this FIXME and removed MacOS from the list. There are clearly some issues with multiprocessing that will need to be debugged properly, but I do not have access to a Mac at the moment. So I propose to just test on Linux and tackle Windows and Mac later.

Comment (David-Araripe) [reply .github/workflows/dev_ci.yml:23]: Looks good to me @martin-sicho.  Indeed, having to debug this complex functionality on the GitHub actions is rather undoable. We could have a look another time

File: .github/workflows/dev_ci.yml [hunk 4]
```diff
@@ -1,82 +1,70 @@
...
       fail-fast: false
     defaults:
       run:
         shell: bash
 
     steps:
-    - uses: actions/checkout@v4
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (David-Araripe) [inline .github/workflows/dev_ci.yml:33]: Just one question, why should we fetch all the history for all tags and branches here?

**Review by David-Araripe — APPROVED** [review]
Ready to merge in opinion. We can have a look at the differing parallel processing behavior later on. I'm not so familiar with the documentation writing/pushing part so I don't have input on that part. It does look ok from the commands I saw!

File: .github/workflows/dev_ci.yml (hunk 4 above)
Comment (martin-sicho) [reply .github/workflows/dev_ci.yml:33]: It is for [setuptools-scm](https://github.com/pypa/setuptools_scm) to label  the version of the package correctly. If tags are missing, the version will not follow the intended versioning scheme. You can see the problem [here](https://github.com/martin-sicho/QSPRpred/actions/runs/7602385899/job/20702887043) (see the *Check Environment Info* step).

Comment (David-Araripe) [reply .github/workflows/dev_ci.yml:33]: Super! I wasn't aware of that. Thanks Martin!

**Pull request merged**